src/rgw/rgw_aio.cc

```diff
@@ -44,6 +61,20 @@ void cb(librados::completion_t, void* arg) {
   s->aio->put(r);
 }
 
+void cache_aio_cb(sigval_t sigval) {
+  D3nL1CacheRequest* c = static_cast<D3nL1CacheRequest*>(sigval.sival_ptr);
+  int status = c->status();
+  if (status == ECANCELED) {
+    c->r->result = -1;
+    c->aio->put(*(c->r));
+    return;
+  } else if (status == 0) {
+    c->finish();
+    c->r->result = 0;
+    c->aio->put(*(c->r));
+  }
+}
```

Comment:
yes, addresses the 3 items

Reply:
I have implemented new `cache_aio_abstract()` and `remote_aio_abstract()` for both Local (L1) and remote cache(L2-via s3) operations, which allow you to have `d->flush() `instead of `d->drain()` without any issue. You can find my changes in the commit down below.

https://github.com/ekaynar/ceph-master/commit/379ab0bd45d925310f51b79c4e5e552992122eaf

Reply:
thank you, switched to `d->flush()`